src/transports/ssh_libssh.c

```diff
@@ -0,0 +1,321 @@
+	}
+
+	if (method & SSH_AUTH_METHOD_INTERACTIVE) {
+		*out |= GIT_CREDENTIAL_SSH_INTERACTIVE;
+	}
+
+	return 0;
+}
+
+int git__ssh_agent_auth(git_ssh_session *s, git_credential_ssh_key *c)
+{
+	int rc = GIT_SSH_ERROR_NONE;
+
+	rc = ssh_userauth_agent(s->session, c->username);
+	if (rc != SSH_OK) {
+
+	}
+
+	rc = ssh_userauth_publickey(s->session, c->username, NULL);
+	if (rc != SSH_OK) {
+
+	}
+
+	return rc;
+}
+
+int _git_ssh_authenticate_session(git_ssh_session *s, git_credential *cred)
+{
+	int rc;
+
+	do {
+		git_error_clear();
+		switch (cred->credtype) {
+			case GIT_CREDENTIAL_USERPASS_PLAINTEXT: {
+				git_credential_userpass_plaintext *c = (git_credential_userpass_plaintext *)cred;
+				rc = ssh_userauth_password(s->session, c->username, c->password);
+				break;
+			}
+			case GIT_CREDENTIAL_SSH_KEY: {
+				git_credential_ssh_key *c = (git_credential_ssh_key *)cred;
+
+				if (c->privatekey) {
+					ssh_key privkey;
+					char *passphrase = c->passphrase ? c->passphrase : "";
+
+					rc = ssh_pki_import_privkey_file(c->privatekey, passphrase, NULL, NULL, &privkey);
+					if (rc != SSH_OK)
+						break;
+
+					rc = ssh_userauth_publickey(s->session, c->username, privkey);
+
+					ssh_key_free(privkey);
+				} else
+					rc = git__ssh_agent_auth(s, c);
+
+				break;
+			}
+			case GIT_CREDENTIAL_SSH_CUSTOM: {
+				git_credential_ssh_custom *c = (git_credential_ssh_custom *)cred;
+
+				/* AFAICS there's no sign callback stuff */
+				GIT_UNUSED(c);
+
+				rc = -1;
+				break;
+			}
+			case GIT_CREDENTIAL_SSH_INTERACTIVE: {
+				git_credential_ssh_interactive *c = (git_credential_ssh_interactive *)cred;
+
+				rc = ssh_userauth_kbdint(s->session, c->username, NULL);
+				if (rc == SSH_AUTH_INFO) {
+					const char *name = ssh_userauth_kbdint_getname(s->session);
+					const char *instr = ssh_userauth_kbdint_getinstruction(s->session);
+					int num_prompts = ssh_userauth_kbdint_getnprompts(s->session);
+					git_credential_ssh_interactive_prompt *prompts;
+					git_credential_ssh_interactive_response *responses;
+					int i = 0;
+
+					prompts = git__calloc(num_prompts, sizeof(*prompts));
+					responses = git__calloc(num_prompts, sizeof(*responses));
+
+					for (i = 0; i < num_prompts; i++) {
+						prompts[i].text = (char *)ssh_userauth_kbdint_getprompt(s->session, i, (char *)&prompts[i].echo);
+						prompts[i].length = strlen(prompts[i].text);
+					}
+
+					c->prompt_callback(s, name, strlen(name),
+									   instr, strlen(instr),
+									   num_prompts, prompts, responses, NULL);
+
+					for (i = 0; i < num_prompts; i++) {
+						ssh_userauth_kbdint_setanswer(s->session, i, responses[i].text);
+					}
+
+					git__free(prompts);
+					git__free(responses);
+				}
+				break;
+			}
+#ifdef GIT_SSH_MEMORY_CREDENTIALS
+			case GIT_CREDTYPE_SSH_MEMORY: {
+				git_cred_ssh_key *c = (git_cred_ssh_key *)cred;
+
+				assert(c->username);
+				assert(c->privatekey);
+
+				rc = libssh2_userauth_publickey_frommemory(
```

Comment:
There are a series of `ssh_pki_import_privkey_` functions provided by libssh which let you specify an in-memory key in different formats. The one that would let us keep the same for both looks like it would be `ssh_pki_import_privkey_base64` as it oddly doesn't seem to allow you just pass the raw data, so we'd have to first convert it into base64.

Reply:
Or it looks like they might mean base64 as in the format you would find on-disk